edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Core
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity.Service;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+
+    /// <summary>
+    /// This is a tree implementation that uses the ParentScopes property
+    /// to reconstruct the hierarchy of modules/devices under the current
+    /// Edge device in a nested Edge environment.
+    /// </summary>
+    class ServiceIdentityTree : IAuthenticationChainProvider
+    {
+        public event EventHandler<string> ServiceIdentityRemoved;
+
+        readonly string rootDeviceId;
+        AtomicReference<ImmutableDictionary<string, ServiceIdentityTreeNode>> nodes;
+
+        public ServiceIdentityTree(string rootDeviceId)
+        {
+            this.rootDeviceId = Preconditions.CheckNonWhiteSpace(rootDeviceId, nameof(rootDeviceId));
+            this.nodes = new AtomicReference<ImmutableDictionary<string, ServiceIdentityTreeNode>>(ImmutableDictionary<string, ServiceIdentityTreeNode>.Empty);
+        }
+
+        public void InsertOrUpdate(ServiceIdentity identity)
+        {
+            // There should always be a valid ServiceIdentity
+            Preconditions.CheckNotNull(identity, nameof(identity));
+
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            var writableSnapshot = new Dictionary<string, ServiceIdentityTreeNode>(snapshot);
+            if (snapshot.ContainsKey(identity.Id))
+            {
+                // Update case - this is just remove + re-insert,
+                // except we don't remove any existing children.
+                // There's no scenario where losing a link to
+                // either the parent or a child wouldn't result
+                // in a call to RemoveRecursive().
+                this.RemoveSingleNode_NoCallback(writableSnapshot, identity.Id);
+            }
+
+            // Insert case
+            if (identity.IsModule)
+            {
+                this.InsertModuleIdentity(writableSnapshot, identity);
+            }
+            else
+            {
+                this.InsertDeviceIdentity(writableSnapshot, identity);
+            }
+
+            // Write back to the atomic field
+            if (!this.nodes.CompareAndSet(snapshot, writableSnapshot.ToImmutableDictionary()))
+            {
+                // There's only ever one thread calling InsertOrUpdate, so this should never fail
+                throw new InvalidOperationException("Multi-threaded insert/update detected on ServiceIdentityTree");
+            }
+        }
+
+        public void Remove(string id)
+        {
+            // In the case of an explicit removal, the device/module is
```

Comment:
should we introduce lock mechanism here rather than in the caller class, DeviceScopeIdentitiesCache?